### ISSUE TITLE
Misc optimization to benchmark and ergonomic ideas

### ISSUE THREAD

**Issue by ariasanovsky** [ariasanovsky/azdopt#14]
## Optimizations

- ~~[ ] [transformers in RL](https://arxiv.org/pdf/2301.03044.pdf)~~
  - relegating to a research task
- ~~[ ] set up benchmarks~~
  - exploring single-threaded search tree
  - compare across different numbers of episodes, epochs, problems, sizes
  - use trivial model
- ~~[ ] `rand` -> `fastrand`~~
  - irrelevant to actual performance -- neural networks are always the bottleneck
- ~~[ ] try different groupings of `(&a, ..., &mut z).into_par_iter()` for performance~~
  - ~~possibly~~ irrelevant to actual performance -- neural networks are likely the bottleneck
- [x] remove `to_vec` from `IRState`
- [x] remove `to_vec` from `State`
  - instead, update the state vector as an outparam (`new` and `act`)
- [x] refactor `Transitions`
  - a lot of unnecessary `Vec` allocates
  - Resolved with Issue #29
- [x] refactor to "exhaust" search space
  - if $as$ is terminal, mark the $(s, a)$ action data so it is not revisited
    - actions are `Exhausted` or `Active`
  - if all actions from $s$ are exhausted, then $s$ is exhausted
    - this may involve removing $s$ from the tree
  - extend "terminal" to "exhausted" recursively to both state-action pairs and to states
- [x] `MuZero` to include reward estimates
- [x] minimize tensor allocations
  - ?gradients: 1 per search instead of 1 per episode
  - lots of excessive array cloning when forming tensors
  - lots of tensor cloning when measuring loss
- [x] grads don't use tape efficiently
  - are the semantics around the Y-shaped architecture in `examples/04-r44.rs` correct?
    - losses decrease for both, but maybe there is misuse/accidentally correct use
  - eliminate the extra call to `forward` if possible
- [x] use `copy_to` when filling arrays from tensors
- [x] use `copy_from` when filling tensors with array data
  - `[[T; M]; N]` has an unstable `flatten` to `&[T]` with `#![feature(slice_flatten)]`
- [x] use a `Vec<BTreeMap<P>>` so we can return a `Vec<&mut Entry>`
  - Resolved with Issue #29
- [x] Wrapper struct for stacks of state nodes to analyze with CNNs
- [x] refactor `IQMinTree` and `examples/04-r44.rs` to match `INTMinTree`
  - this task is a relocation of a task from #21 
- [x] Improve `INTStateData`
  - [x] we only need to sort the visited actions during each update
  - [x] use a `VecDeque` and take care to prevent fragmentation
    - maybe a manually managed `Vec` which is a sorted buffer with 3 pieces
      - **exhausted**: the actions at the back which only lead to exhausted states
      - **visited**: the next actions from the back, sorted by $u(s, a)$
      - **unvisited**: the remaining actions, sorted by $p(s, a)$ (or $u(s, a)$ if we have $r(s, a)$)
- [x] allow user-specified functions for $u(s, a)$, or equivalent methods for selecting actions

There are 2 or 3 stagnation thresholds:
- [x] to reset `time`
  - I still like the idea of this resetting every epoch, but why not let it be adjustable
- [x] to reset `prohibited actions`
  - we are running with the hypothesis that keeping a running list of prohibited actions is useful for training data
  - but in order to gather this training data, we cannot 
- [x] to reset the `State` underlying the the `StateNode<State>` (or whatever higher-level wrapper we make

This may require fixing the `CostLog` state machine a little. It seems to be working, though. It's good to pass a `&mut s: State` to mutate over a batch rather than overloading `CostLog`'s API with a slew of different reset methods this early in the API.
## Ergonomic ideas

# ?Factor `Cost` out to maintain API flexibility
  - [x] Why not just use a `let cost_eval: impl Fn(&C) -> f32 + Send/Sync = |c: &C| { let f = todo!(); c }` instead of hard-coding all these trait bounds?

## Data structures

- [ConstLru](https://docs.rs/const-lru/latest/const_lru/struct.ConstLru.html) or [cached](https://crates.io/crates/cached)
- any alternatives to the [priority_queue](https://docs.rs/priority-queue/latest/priority_queue/)?
- handroll a MCTS tree? check out other crates?

## Questions

- How to limit allocates in the tree's set of keys?
  - When a each state (relative to the root) is associated a value in the tree, a vector of length equal to the number of actions
  - Dozen of trees, thousands of states in each, hundreds (soon thousands) of actions possible from each
  - We could limit the capacity of the vector permanently with fixed-size vectors
  - E.g., store only the best $100$, by upper estimate (monotone in reward and probability)
  - Alternatively, store a small number of actions and expand the list after a threshold is reached
- How to limit allocations in the `Transitions` data?
  - `Transitions` holds a `Vec` of `Vec`'s (the action data)
  - Refactor to reuse the previously allocated vectors
  - Use an arena
  - (!?) Refactor to use only the `Vec<&mut IQStateData>`
    - the borrow-checker doesn't know that the paths are distinct and denies this in `safe` code

**Comment by ariasanovsky** [comment]
> refactor to "exhaust" search space 

and

> refactor Transitions

can be solve with the following workflow. Replace `Vec<Vec<Key>>` with `Vec<Entry>`.

![Episode Logging Workflow](https://github.com/ariasanovsky/azdopt/assets/14944402/19c7d080-9698-4750-965e-83a9db3aa327)

**Comment by ariasanovsky** [comment]
![Epoch Logging Workflow](https://github.com/ariasanovsky/azdopt/assets/14944402/42e46aef-928b-4990-a330-55579a5eb4d3)

**Comment by ariasanovsky** [comment]
> ![Epoch Logging Workflow](https://user-images.githubusercontent.com/14944402/281116527-42e46aef-928b-4990-a330-55579a5eb4d3.png)

**Comment by ariasanovsky** [comment]
# ?Factor `Cost` out to maintain API flexibility
- [ ] Why not just use a `let cost_eval: impl Fn(&C) -> f32 + Send/Sync = |c: &C| { let f = todo!(); c }` instead of hard-coding all these trait bounds?

**Comment by ariasanovsky** [comment]
Moved the misplaced optimization from #24 here.

**Comment by ariasanovsky** [comment]
Closed with #73 #74
